openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,551 @@
+		return "", fmt.Errorf("failed to parse %q annotation value: %v", workloadAnnotationKey, err)
+	}
+
+	effect, ok := managementAnnotationValue[podWorkloadAnnotationEffect]
+	if !ok {
+		return "", fmt.Errorf("the workload annotation value %q does not have %q key", workloadAnnotationKey, podWorkloadAnnotationEffect)
+	}
+	return effect, nil
+}
+
+func stripResourcesAnnotations(annotations map[string]string) {
+	for k := range annotations {
+		if strings.HasPrefix(k, containerResourcesAnnotationPrefix) {
+			delete(annotations, k)
+		}
+	}
+}
+
+func stripWorkloadAnnotations(annotations map[string]string) {
+	for k := range annotations {
+		if strings.HasPrefix(k, podWorkloadAnnotationPrefix) {
+			delete(annotations, k)
+		}
+	}
+}
+
+func addWorkloadAnnotations(annotations map[string]string, workloadType string) {
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	annotations[workloadAnnotation] = fmt.Sprintf(`{"%s":"%s"}`, podWorkloadAnnotationEffect, workloadEffectPreferredDuringScheduling)
+}
+
+func (a *managementCPUsOverride) Validate(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) (err error) {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	ns, err := a.getPodNamespace(attr)
+	if err != nil {
+		return err
+	}
+
+	var allErrs field.ErrorList
+	workloadType, err := getWorkloadType(pod.Annotations)
+	if err != nil {
+		allErrs = append(allErrs, getPodInvalidWorkloadAnnotationError(pod.Annotations, err.Error()))
+	}
+
+	cpusetAnnotationKey := fmt.Sprintf("%s.%s.cpuset", containerResourcesAnnotationPrefix, workloadType)
+	for k := range pod.Annotations {
+		// the pod should not have any cpuset annotation
+		if strings.HasPrefix(k, cpusetAnnotationKey) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("metadata.annotations"), pod.Annotations, "the pod should not have any cpuset resource annotation"))
+		}
+	}
+
+	// the pod does not have workload annotation
+	if len(workloadType) == 0 {
+		for k := range pod.Annotations {
+			// pod has resource annotation
+			if strings.HasPrefix(k, containerResourcesAnnotationPrefix) {
+				allErrs = append(allErrs, getPodInvalidWorkloadAnnotationError(pod.Annotations, fmt.Sprintf("the pod without workload annotation can not have resource annotation %q", k)))
+			}
+		}
+
+		allContainers := append([]coreapi.Container{}, pod.Spec.InitContainers...)
+		allContainers = append(allContainers, pod.Spec.Containers...)
+		for _, c := range allContainers {
+			for r := range c.Resources.Requests {
+				resourceName := string(r)
+				if strings.HasSuffix(resourceName, fmt.Sprintf("%scores", podWorkloadAnnotationPrefix)) {
```

Comment:
HasSuffix is unusual.  Do we know the exact name?

Reply:
the resource name can be different by workload type, so in the case when I do not have any workload type specified under the pod annotation we should not have any new workload resource under containers resources or limits.